torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -14,6 +14,25 @@ class AllGatherResult(NamedTuple):
     all_gather_input_numels: List[int]
 
 
+class AllGatherState(NamedTuple):
+    all_gather_result: AllGatherResult
+    event: torch.cuda.Event  # copy-out
+
+
+class AllGatherStateHolder:
+    def __init__(self):
+        self._state: Optional[AllGatherState] = None
+
+    def put(self, state: AllGatherState) -> None:
+        assert self._state is None, "Expects to hold only one all-gather state"
+        self._state = state
+
+    def pop(self) -> Optional[AllGatherState]:
+        state = self._state
+        self._state = None
+        return state
+
+
```

Comment:
Is there an indirect synchronization that ensures that the first all-gather happens after the optimizer step for the corresponding parameters completes?

Reply:
Sorry, since this PR does not cover forward/backward training, this synchronization step is missing in this PR.

It will be done in the first forward to run (almost always the root's forward).
https://github.com/pytorch/pytorch/pull/118004/files#diff-a2be7eb5a3d826e913056d5cf851e94be0914e8cca00dfb1a820fbae747eb494R73